File: snippets/screens/swiftui/permissions_screen/2/permissions_screen.swift
import SwiftUI

struct PermissionsScreen: View {
    @State private var permissions: [PermissionItem] = [
        PermissionItem(
            icon: "bell.fill",
            title: "Notifications",
            description: "Get notified about new challenges and your progress",
            isRequired: false,
            isGranted: false
        ),
        PermissionItem(
            icon: "camera.fill",
            title: "Camera",
            description: "Take photos for your profile and share achievements",
            isRequired: false,
            isGranted: false
        ),
        PermissionItem(
            icon: "mic.fill",
            title: "Microphone",
            description: "Use voice commands for hands-free navigation",
            isRequired: false,
            isGranted: false
        ),
        PermissionItem(
            icon: "location.fill",
            title: "Location",
            description: "Find nearby users and location-based challenges",
            isRequired: false,
            isGranted: false
        )
    ]

    @State private var contentOpacity: Double = 0
    @State private var contentOffset: CGFloat = 50

    var body: some View {
        VStack {
            // Header
            VStack(alignment: .leading, spacing: 24) {
                HStack(spacing: 16) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 12)
                            .fill(Color(red: 0.39, green: 0.40, blue: 0.95).opacity(0.1))
                            .frame(width: 48, height: 48)

                        Image(systemName: "shield.checkered")
                            .font(.system(size: 24))
                            .foregroundColor(Color(red: 0.39, green: 0.40, blue: 0.95))
                    }

                    VStack(alignment: .leading, spacing: 4) {
                        Text("Permissions")
                            .font(.system(size: 24, weight: .bold))
                            .foregroundColor(Color(red: 0.12, green: 0.16, blue: 0.22))

                        Text("Help us personalize your experience")
                            .font(.system(size: 14))
                            .foregroundColor(Color(red: 0.42, green: 0.45, blue: 0.50))
                    }

                    Spacer()
                }

                Text("We'd like your permission to access the following features to enhance your AIQ experience:")
                    .font(.system(size: 16))
                    .foregroundColor(Color(red: 0.42, green: 0.45, blue: 0.50))
                    .lineSpacing(4)
            }
            .padding(.horizontal, 24)
            .padding(.top, 20)

            // Permissions list
            ScrollView {
                LazyVStack(spacing: 16) {
                    ForEach(Array(permissions.enumerated()), id: \.offset) { index, permission in
                        PermissionCard(
                            permission: permission,
                            onToggle: { togglePermission(at: index) }
                        )
                    }
                }
                .padding(.horizontal, 24)
                .padding(.top, 32)
            }

            Spacer()

            // Action buttons
            VStack(spacing: 12) {
                Button(action: continueToApp) {
                    Text("Continue")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .frame(height: 56)
                        .background(Color(red: 0.39, green: 0.40, blue: 0.95))
                        .cornerRadius(16)
                }

                Button(action: skipPermissions) {
                    Text("Skip for now")
                        .font(.system(size: 16, weight: .medium))
                        .foregroundColor(Color(red: 0.42, green: 0.45, blue: 0.50))
                        .frame(height: 44)
                }
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 24)
        }
        .opacity(contentOpacity)
        .offset(y: contentOffset)
        .background(Color.white)
        .onAppear {
            startAnimations()
        }
    }

    private func startAnimations() {
        withAnimation(.easeOut(duration: 0.8)) {
            contentOpacity = 1
            contentOffset = 0
        }
    }

    private func togglePermission(at index: Int) {
        withAnimation(.easeInOut(duration: 0.2)) {
            permissions[index].isGranted.toggle()
        }

        // Simulate permission request delay
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            // Show feedback if needed
        }
    }

    private func continueToApp() {
        // Navigate to interests screen
        // NavigationManager.shared.navigate(to: .interests)
    }

    private func skipPermissions() {
        // Navigate to interests screen
        // NavigationManager.shared.navigate(to: .interests)
    }
}

struct PermissionCard: View {
    let permission: PermissionItem
    let onToggle: () -> Void

    var body: some View {
        HStack(spacing: 16) {
            // Icon
            ZStack {
                RoundedRectangle(cornerRadius: 12)
                    .fill(permission.isGranted ?
                          Color(red: 0.06, green: 0.73, blue: 0.51).opacity(0.1) :
                          Color(red: 0.42, green: 0.45, blue: 0.50).opacity(0.1)
                    )
                    .frame(width: 48, height: 48)

                Image(systemName: permission.icon)
                    .font(.system(size: 20))
                    .foregroundColor(permission.isGranted ?
                                   Color(red: 0.06, green: 0.73, blue: 0.51) :
                                   Color(red: 0.42, green: 0.45, blue: 0.50)
                    )
            }

            // Content
            VStack(alignment: .leading, spacing: 4) {
                HStack {
                    Text(permission.title)
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(Color(red: 0.12, green: 0.16, blue: 0.22))

                    if permission.isRequired {
                        Text("Required")
                            .font(.system(size: 10, weight: .medium))
                            .foregroundColor(.white)
                            .padding(.horizontal, 6)
                            .padding(.vertical, 2)
                            .background(Color(red: 0.94, green: 0.27, blue: 0.27))
                            .cornerRadius(8)
                    }

                    Spacer()
                }

                Text(permission.description)
                    .font(.system(size: 14))
                    .foregroundColor(Color(red: 0.42, green: 0.45, blue: 0.50))
                    .lineSpacing(2)
            }

            // Toggle button
            Button(action: onToggle) {
                Text(permission.isGranted ? "Granted" : "Allow")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundColor(.white)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 8)
                    .background(permission.isGranted ?
                              Color(red: 0.06, green: 0.73, blue: 0.51) :
                              Color(red: 0.39, green: 0.40, blue: 0.95)
                    )
                    .cornerRadius(8)
            }
        }
        .padding(16)
        .background(Color.white)
        .overlay(
            RoundedRectangle(cornerRadius: 16)
                .stroke(Color(red: 0.90, green: 0.91, blue: 0.92), lineWidth: 1)
        )
        .shadow(color: .black.opacity(0.05), radius: 10, x: 0, y: 2)
    }
}

struct PermissionItem {
    let icon: String
    let title: String
    let description: String
    let isRequired: Bool
    var isGranted: Bool
}

#Preview {
    PermissionsScreen()
}

File: snippets/screens/flutter/permissions_screen/2/permissions_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class PermissionsScreen extends StatefulWidget {
  const PermissionsScreen({super.key});

  @override
  State<PermissionsScreen> createState() => _PermissionsScreenState();
}

class _PermissionsScreenState extends State<PermissionsScreen>
    with TickerProviderStateMixin {
  late AnimationController _animationController;
  late Animation<double> _slideAnimation;
  late Animation<double> _fadeAnimation;

  final List<PermissionItem> _permissions = [
    PermissionItem(
      icon: Icons.notifications,
      title: 'Notifications',
      description: 'Get notified about new challenges and your progress',
      isRequired: false,
      isGranted: false,
    ),
    PermissionItem(
      icon: Icons.camera_alt,
      title: 'Camera',
      description: 'Take photos for your profile and share achievements',
      isRequired: false,
      isGranted: false,
    ),
    PermissionItem(
      icon: Icons.mic,
      title: 'Microphone',
      description: 'Use voice commands for hands-free navigation',
      isRequired: false,
      isGranted: false,
    ),
    PermissionItem(
      icon: Icons.location_on,
      title: 'Location',
      description: 'Find nearby users and location-based challenges',
      isRequired: false,
      isGranted: false,
    ),
  ];

  @override
  void initState() {
    super.initState();

    SystemChrome.setSystemUIOverlayStyle(
      const SystemUiOverlayStyle(
        statusBarColor: Colors.transparent,
        statusBarIconBrightness: Brightness.dark,
      ),
    );

    _animationController = AnimationController(
      duration: const Duration(milliseconds: 800),
      vsync: this,
    );

    _slideAnimation = Tween<double>(
      begin: 1.0,
      end: 0.0,
    ).animate(CurvedAnimation(
      parent: _animationController,
      curve: Curves.easeOutCubic,
    ));

    _fadeAnimation = Tween<double>(
      begin: 0.0,
      end: 1.0,
    ).animate(CurvedAnimation(
      parent: _animationController,
      curve: Curves.easeIn,
    ));

    _animationController.forward();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void _requestPermission(int index) async {
    // Simulate permission request
    await Future.delayed(const Duration(milliseconds: 500));

    setState(() {
      _permissions[index].isGranted = !_permissions[index].isGranted;
    });

    // Show feedback
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(
        content: Text(
          _permissions[index].isGranted
              ? '${_permissions[index].title} permission granted'
              : '${_permissions[index].title} permission denied',
        ),
        backgroundColor: _permissions[index].isGranted
            ? Colors.green
            : Colors.orange,
        duration: const Duration(seconds: 2),
      ),
    );
  }

  void _continueToApp() {
    Navigator.of(context).pushReplacementNamed('/interests');
  }

  void _skipPermissions() {
    Navigator.of(context).pushReplacementNamed('/interests');
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: AnimatedBuilder(
          animation: _animationController,
          builder: (context, child) {
            return Transform.translate(
              offset: Offset(0, 50 * _slideAnimation.value),
              child: FadeTransition(
                opacity: _fadeAnimation,
                child: Padding(
                  padding: const EdgeInsets.all(24.0),
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      const SizedBox(height: 20),
                      // Header
                      Row(
                        children: [
                          Container(
                            width: 48,
                            height: 48,
                            decoration: BoxDecoration(
                              color: const Color(0xFF6366F1).withOpacity(0.1),
                              borderRadius: BorderRadius.circular(12),
                            ),
                            child: const Icon(
                              Icons.security,
                              color: Color(0xFF6366F1),
                              size: 24,
                            ),
                          ),
                          const SizedBox(width: 16),
                          Expanded(
                            child: Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: const [
                                Text(
                                  'Permissions',
                                  style: TextStyle(
                                    fontSize: 24,
                                    fontWeight: FontWeight.bold,
                                    color: Color(0xFF1F2937),
                                  ),
                                ),
                                Text(
                                  'Help us personalize your experience',
                                  style: TextStyle(
                                    fontSize: 14,
                                    color: Color(0xFF6B7280),
                                  ),
                                ),
                              ],
                            ),
                          ),
                        ],
                      ),
                      const SizedBox(height: 32),
                      const Text(
                        'We\'d like your permission to access the following features to enhance your AIQ experience:',
                        style: TextStyle(
                          fontSize: 16,
                          color: Color(0xFF6B7280),
                          height: 1.5,
                        ),
                      ),
                      const SizedBox(height: 32),
                      // Permissions list
                      Expanded(
                        child: ListView.builder(
                          itemCount: _permissions.length,
                          itemBuilder: (context, index) {
                            final permission = _permissions[index];
                            return Container(
                              margin: const EdgeInsets.only(bottom: 16),
                              decoration: BoxDecoration(
                                color: Colors.white,
                                borderRadius: BorderRadius.circular(16),
                                border: Border.all(
                                  color: const Color(0xFFE5E7EB),
                                  width: 1,
                                ),
                                boxShadow: [
                                  BoxShadow(
                                    color: Colors.black.withOpacity(0.05),
                                    blurRadius: 10,
                                    offset: const Offset(0, 2),
                                  ),
                                ],
                              ),
                              child: ListTile(
                                contentPadding: const EdgeInsets.all(16),
                                leading: Container(
                                  width: 48,
                                  height: 48,
                                  decoration: BoxDecoration(
                                    color: permission.isGranted
                                        ? const Color(0xFF10B981).withOpacity(0.1)
                                        : const Color(0xFF6B7280).withOpacity(0.1),
                                    borderRadius: BorderRadius.circular(12),
                                  ),
                                  child: Icon(
                                    permission.icon,
                                    color: permission.isGranted
                                        ? const Color(0xFF10B981)
                                        : const Color(0xFF6B7280),
                                  ),
                                ),
                                title: Row(
                                  children: [
                                    Text(
                                      permission.title,
                                      style: const TextStyle(
                                        fontSize: 16,
                                        fontWeight: FontWeight.w600,
                                        color: Color(0xFF1F2937),
                                      ),
                                    ),
                                    if (permission.isRequired) ...[
                                      const SizedBox(width: 8),
                                      Container(
                                        padding: const EdgeInsets.symmetric(
                                          horizontal: 6,
                                          vertical: 2,
                                        ),
                                        decoration: BoxDecoration(
                                          color: const Color(0xFFEF4444),
                                          borderRadius: BorderRadius.circular(8),
                                        ),
                                        child: const Text(
                                          'Required',
                                          style: TextStyle(
                                            fontSize: 10,
                                            color: Colors.white,
                                            fontWeight: FontWeight.w500,
                                          ),
                                        ),
                                      ),
                                    ],
                                  ],
                                ),
                                subtitle: Padding(
                                  padding: const EdgeInsets.only(top: 4),
                                  child: Text(
                                    permission.description,
                                    style: const TextStyle(
                                      fontSize: 14,
                                      color: Color(0xFF6B7280),
                                      height: 1.3,
                                    ),
                                  ),
                                ),
                                trailing: SizedBox(
                                  width: 80,
                                  child: ElevatedButton(
                                    onPressed: () => _requestPermission(index),
                                    style: ElevatedButton.styleFrom(
                                      backgroundColor: permission.isGranted
                                          ? const Color(0xFF10B981)
                                          : const Color(0xFF6366F1),
                                      foregroundColor: Colors.white,
                                      padding: const EdgeInsets.symmetric(
                                        horizontal: 12,
                                        vertical: 8,
                                      ),
                                      shape: RoundedRectangleBorder(
                                        borderRadius: BorderRadius.circular(8),
                                      ),
                                    ),
                                    child: Text(
                                      permission.isGranted ? 'Granted' : 'Allow',
                                      style: const TextStyle(
                                        fontSize: 12,
                                        fontWeight: FontWeight.w600,
                                      ),
                                    ),
                                  ),
                                ),
                              ),
                            );
                          },
                        ),
                      ),
                      // Action buttons
                      const SizedBox(height: 24),
                      Column(
                        children: [
                          SizedBox(
                            width: double.infinity,
                            height: 56,
                            child: ElevatedButton(
                              onPressed: _continueToApp,
                              style: ElevatedButton.styleFrom(
                                backgroundColor: const Color(0xFF6366F1),
                                foregroundColor: Colors.white,
                                shape: RoundedRectangleBorder(
                                  borderRadius: BorderRadius.circular(16),
                                ),
                                elevation: 0,
                              ),
                              child: const Text(
                                'Continue',
                                style: TextStyle(
                                  fontSize: 16,
                                  fontWeight: FontWeight.w600,
                                ),
                              ),
                            ),
                          ),
                          const SizedBox(height: 12),
                          TextButton(
                            onPressed: _skipPermissions,
                            child: const Text(
                              'Skip for now',
                              style: TextStyle(
                                fontSize: 16,
                                color: Color(0xFF6B7280),
                                fontWeight: FontWeight.w500,
                              ),
                            ),
                          ),
                        ],
                      ),
                    ],
                  ),
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}

class PermissionItem {
  final IconData icon;
  final String title;
  final String description;
  final bool isRequired;
  bool isGranted;

  PermissionItem({
    required this.icon,
    required this.title,
    required this.description,
    required this.isRequired,
    required this.isGranted,
  });
}